eclass/databases.eclass

```diff
@@ -0,0 +1,505 @@
+			;;
+		mysql)
+			echo "virtual/mysql[server]"
+			;;
+		postgres)
+			echo "dev-db/postgresql[server]"
+			;;
+		redis)
+			echo "dev-db/redis"
+			;;
+		*)
+			die "${ECLASS}: unknown database: ${srvname}"
+	esac
+}
+
+# @FUNCTION: _databases_die
+# @USAGE: <funcname> [msg]
+# @INTERNAL
+# @DESCRIPTION:
+# Print the given message and the path to the server's log file to the console
+# and die.
+#
+# This function supports being called via "nonfatal".
+_databases_die() {
+	local funcname=${1?}
+	shift
+
+	eerror "See the server log for details:"
+	eerror "	$(${funcname} --get-logfile)"
+	die -n "${@}"
+}
+
+# @FUNCTION: _databases_mkdirs
+# @USAGE: <funcname>
+# @INTERNAL
+# @DESCRIPTION:
+# Create all directories that might be needed at runtime.
+_databases_mkdirs() {
+	debug-print-function "${FUNCNAME}" "${@}"
+
+	local funcname=${1?}
+
+	# directories
+	local dbpath=$(${funcname} --get-dbpath)
+	local sockdir=$(${funcname} --get-sockdir)
+	mkdir -p "${dbpath}" || die "Creating database directory failed"
+	mkdir -p "${sockdir}" || die "Creating sockets directory failed"
+
+	# files
+	local logfile_dir=$(dirname "$(${funcname} --get-logfile)")
+	local pidfile_dir=$(dirname "$(${funcname} --get-pidfile)")
+	local sockfile_dir=$(dirname "$(${funcname} --get-sockfile)")
+	mkdir -p "${logfile_dir}" || die "Creating log directory failed"
+	mkdir -p "${pidfile_dir}" || die "Creating pidfile directory filed"
+	mkdir -p "${sockfile_dir}" || die "Creating sockets directory filed"
+}
+
+# @FUNCTION: _databases_stop_service
+# @USAGE: <funcname>
+# @INTERNAL
+# @DESCRIPTION:
+# Default function to stop servers.  Reads PID from a file and sends the TERM
+# signal.
+_databases_stop_service() {
+	debug-print-function "${FUNCNAME}" "${@}"
+
+	local funcname=${1?}
+	local srvname=${funcname:1}
+	local pidfile="$(${funcname} --get-pidfile)"
+
+	ebegin "Stopping ${srvname}"
+	kill "$(<"${pidfile}")"
+	eend $? || ${funcname} --die "Stopping ${srvname} failed"
+}
+
+# @FUNCTION: _databases_dispatch
+# @USAGE: <funcname> <cmd> [args...]
+# @INTERNAL
+# @DESCRIPTION:
+# Process the given command with its options.
+#
+# If "--start" command is used, `_${funcname}_start` function must be defined.
+# Note that directories will be created automatically.
+#
+# If `_${funcname}_stop` function is not declared, the internal
+# `_databases_stop_service` function will be used instead.
+#
+# No `--get` function can be overloaded.
+_databases_dispatch() {
+	local funcname=${1?}
+	local cmd=${2?}
+	shift; shift
+
+	case ${cmd} in
+		--die)
+			_databases_die ${funcname} "${@}"
+			;;
+		--get-depend)
+			_databases_gen_depend ${funcname}
+			;;
+		--get-dbpath)
+			echo "${T}"/${funcname}/db/
+			;;
+		--get-logfile)
+			echo "${T}"/${funcname}/log/${funcname}.log
+			;;
+		--get-pidfile)
+			echo "${T}"/${funcname}/run/${funcname}.pid
+			;;
+		--get-sockdir)
+			echo "${T}"/${funcname}/run/
+			;;
+		--get-sockfile)
+			echo "${T}"/${funcname}/run/${funcname}.sock
+			;;
```

Comment:
Still, do you need the `log` and `run` subdirs? It feels like overkill.

Reply:
OK, removed `log` and `run` from paths